Move logic of network usage monitoring to NetworkProcess

basuke opened the pull request:
#### 65e37ab3fc65158cc79798546f7aa56753112d0a
<pre>
Move logic of network usage monitoring to NetworkProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=287120">https://bugs.webkit.org/show_bug.cgi?id=287120</a>
<a href="https://rdar.apple.com/144241587">rdar://144241587</a>

Reviewed by Ben Nham.

Currently the source of network usage for ResourceMonitor is in ResourceLoader in Web Process. We need to move
this into Network Process to improve the quality of monitoring which allows:

1. Avoid counting the resources comes from the cache.
2. Avoid counting blob and data urls.
3. Not the decoded size after network load but the actual bytes before decoding. (i.e. gzip a.k.a deflate)

To achieve this, a new parameter `bytesTransferredOverNetwork` is added to WebResourceLoader message and Network
Process try to send correct network usage of the request. This number is still decoded buffer size in this patch
and will be implemented in the following patch.

* LayoutTests/http/tests/iframe-monitor/blob-resource-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/blob-resource.html: Copied from LayoutTests/http/tests/iframe-monitor/iframe-unload.html.
* LayoutTests/http/tests/iframe-monitor/cached-resource-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/cached-resource.html: Added.
* LayoutTests/http/tests/iframe-monitor/dark-mode.html:
* LayoutTests/http/tests/iframe-monitor/data-url-resource-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/data-url-resource.html: Copied from LayoutTests/http/tests/iframe-monitor/iframe-unload.html.
* LayoutTests/http/tests/iframe-monitor/eligibility.html:
* LayoutTests/http/tests/iframe-monitor/iframe-unload.html:
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/blob.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/cached.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/data-url.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe.html: Renamed from LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--.html.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe2.html: Renamed from LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--2.html.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe3.html: Renamed from LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--3.html.
* LayoutTests/http/tests/iframe-monitor/resources/generate-byte.py:
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js:
(async setup):
* LayoutTests/http/tests/iframe-monitor/throttler.html:
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::didReceiveBuffer):
* Source/WebCore/loader/ResourceLoader.h:
* Source/WebCore/loader/ResourceMonitor.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::calculateBytesTransferredOverNetworkDelta):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
(WebKit::NetworkDataTask::totalBytesTransferredOverNetwork const):
(WebKit::NetworkDataTask::setTotalBytesTransferredOverNetwork):
* Source/WebKit/NetworkProcess/NetworkLoad.cpp:
(WebKit::NetworkLoad::bytesTransferredOverNetworkDelta const):
* Source/WebKit/NetworkProcess/NetworkLoad.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::bufferingTimerFired):
(WebKit::NetworkResourceLoader::sendBuffer):
(WebKit::NetworkResourceLoader::dataReceivedThroughContentFilter):
(WebKit::NetworkResourceLoader::bytesTransferredOverNetworkDelta const):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::didReceiveData):
(WebKit::ServiceWorkerFetchTask::didReceiveDataFromPreloader):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::didReceiveData):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::curlDidReceiveData):
* Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp:
(WebKit::NetworkDataTaskSoup::didRead):
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::didReceiveData):
* Source/WebKit/WebProcess/Network/WebResourceLoader.h:
* Source/WebKit/WebProcess/Network/WebResourceLoader.messages.in:

Canonical link: <a href="https://commits.webkit.org/290164@main">https://commits.webkit.org/290164@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67d3c301ac1ac1e5c419ca9ac88655464b50522e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8627 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94081 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39864 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91154 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9014 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16814 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68655 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26327 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6899 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80851 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49016 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6648 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35230 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38971 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77011 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36212 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95916 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16283 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11923 "Found 5 new test failures: http/tests/webarchive/cross-origin-stylesheet-crash.html http/tests/webarchive/test-css-url-encoding-shift-jis.html http/tests/webarchive/test-css-url-encoding-utf-8.html http/tests/webarchive/test-css-url-encoding.html http/tests/webarchive/test-preload-resources.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77532 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16539 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76641 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76823 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18948 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21223 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19825 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9375 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16297 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21608 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16038 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19489 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17819 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -2129,6 +2134,7 @@ void NetworkResourceLoader::serviceWorkerDidFinish()
 void NetworkResourceLoader::dataReceivedThroughContentFilter(const SharedBuffer& buffer, size_t encodedDataLength)
 {
+    updateNetworkUsage();
     send(Messages::WebResourceLoader::DidReceiveData(IPC::SharedBufferReference(buffer), encodedDataLength));
```
I haven't had a chance to fully read through this yet, but my very initial thought was that instead of adding a new IPC type, would it make more sense to just change `DidReceiveData` to include both the encoded data length (which I think is post-compression) as well as the delta of the network bytes transferred so far?

basuke replied:
Actually, `DidReceiveData` already expects the encodedSize if I read the parameter names correctly. 
```
    DidReceiveData(IPC::SharedBufferReference data, uint64_t encodedDataLength)
```
But the size of buffer is just sent to the web process. I've hesitated to add extra parameter name to there. And I also not sure I should change the current implementation without knowing other platform's usage. Also this value `encodedDataLength` is passed to the web process for both cached resource and blob.

If we change this, I think this is the good shape of this message:
```
    DidReceiveData(IPC::SharedBufferReference data, uint64_t dataLength, unint64_t networkUsage)
```
This is more obvious than current shape. Or not changing the shape and simply change the implementation of the `encodedDataLength` to much the name.

basuke replied:
In latter case, we should send zero to cached response and blob, data url access.

basuke replied:
As far as the test results by correcting the meaning of second parameter `encodedDataLength`, it seems no issue in the result of layout test with the change on a different draft PR for testing purpose https://github.com/WebKit/WebKit/pull/40105#issuecomment-2638905347 so we can go with this change.

basuke replied:
Ah, no. This PR just move the logic to implement platform specific calculations. Let me push the one with the actual change first.

bnham commented on Source/WebKit/NetworkProcess/NetworkDataTask.h:
```diff
@@ -162,6 +164,7 @@ class NetworkDataTask : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<N
     void scheduleFailure(FailureType);
 
     void restrictRequestReferrerToOriginIfNeeded(WebCore::ResourceRequest&);
+    void updateNetworkUsage(size_t usage) { m_networkUsage = usage; }
```
nit: seems like this should just be named `setNetworkUsage`.

"NetworkUsage" also seems pretty vague to me--maybe "bytesTransferredOverNetwork" or something?

bnham reviewed (approved):
It seems like it'd be cleaner to change the `WebResourceLoader::DidReceiveData` IPC to include the number of bytes transferred over the network associated with that chunk of data, instead of adding a new IPC. We might have to change this for perf reasons as the power bot may detect the extra unnecessary IPCs from this patch as a power regression. But I don't think it affects the correctness of the patch so I'll approve it.

bnham commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -2181,6 +2182,18 @@ void NetworkResourceLoader::useRedirectionForCurrentNavigation(WebCore::Resource
     m_redirectionForCurrentNavigation = makeUnique<WebCore::ResourceResponse>(WTFMove(response));
 }
 
+uint64_t NetworkResourceLoader::bytesTransferredOverNetwork()
```
This name feels misleading, since it is really "total bytes transferred over the network since the last time you called this function", rather than "total bytes transferred over the network".

I think I would either rename this method to reflect that (`bytesTransferredOverNetworkDelta?`) or refactor to make this clearer.

bnham commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -397,6 +397,8 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
 {
     WTFEmitSignpost(m_task.get(), DataTask, "received %zd bytes", data.size());
 
+    setTotalBytesTransferredOverNetwork(totalBytesTransferredOverNetwork() + data.size());
```
As you noted, this is going to be reporting decompressed size rather than compressed size. Please file a follow up radar to fix this during convergence.

cdumez commented on Source/WebKit/NetworkProcess/NetworkDataTask.h:
```diff
@@ -176,10 +180,22 @@ class NetworkDataTask : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<N
     bool m_failureScheduled { false };
 };
 
+size_t NetworkDataTask::bytesTransferredOverNetworkDelta()
```
I don't think this needs to be inline. We should probably just move it to the cpp file to avoid polluting the header.

basuke replied:
But then in `NetworkLoad::bytesTransferredOverNetworkDelta() const` we need to ref `m_task`.

cdumez replied:
I don't understand what your comment has to do with making the function inline.

cdumez commented on Source/WebKit/NetworkProcess/NetworkDataTask.h:
```diff
@@ -150,6 +150,9 @@ class NetworkDataTask : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<N
     virtual void setTimingAllowFailedFlag() { }
 
+    inline size_t bytesTransferredOverNetworkDelta();
+    size_t totalBytesTransferredOverNetwork() { return m_totalBytesTransferredOverNetwork; }
```
should be `const`

basuke replied:
It changes the member variable so it cannot be `const`.

basuke replied:
I'll change the name to `calculateBytesTransferredOverNetworkDelta` to make not look like getter.

cdumez replied:
Are you commenting on the wrong line? The line I commented on looks like a trivial getter than CAN be const.

basuke committed:
Move logic of network usage monitoring to NetworkProcess
https://bugs.webkit.org/show_bug.cgi?id=287120
rdar://144241587

Reviewed by Ben Nham.

Currently the source of network usage for ResourceMonitor is in ResourceLoader in Web Process. We need to move
this into Network Process to improve the quality of monitoring which allows:

1. Avoid counting the resources comes from the cache.
2. Avoid counting blob and data urls.
3. Not the decoded size after network load but the actual bytes before decoding. (i.e. gzip a.k.a deflate)

To achieve this, a new parameter `bytesTransferredOverNetwork` is added to WebResourceLoader message and Network
Process try to send correct network usage of the request. This number is still decoded buffer size in this patch
and will be implemented in the following patch.

* LayoutTests/http/tests/iframe-monitor/blob-resource-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/blob-resource.html: Copied from LayoutTests/http/tests/iframe-monitor/iframe-unload.html.
* LayoutTests/http/tests/iframe-monitor/cached-resource-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/cached-resource.html: Added.
* LayoutTests/http/tests/iframe-monitor/dark-mode.html:
* LayoutTests/http/tests/iframe-monitor/data-url-resource-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/data-url-resource.html: Copied from LayoutTests/http/tests/iframe-monitor/iframe-unload.html.
* LayoutTests/http/tests/iframe-monitor/eligibility.html:
* LayoutTests/http/tests/iframe-monitor/iframe-unload.html:
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/blob.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/cached.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/data-url.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe.html: Renamed from LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--.html.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe2.html: Renamed from LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--2.html.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe3.html: Renamed from LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--3.html.
* LayoutTests/http/tests/iframe-monitor/resources/generate-byte.py:
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js:
(async setup):
* LayoutTests/http/tests/iframe-monitor/throttler.html:
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::didReceiveBuffer):
* Source/WebCore/loader/ResourceLoader.h:
* Source/WebCore/loader/ResourceMonitor.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::calculateBytesTransferredOverNetworkDelta):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
(WebKit::NetworkDataTask::totalBytesTransferredOverNetwork const):
(WebKit::NetworkDataTask::setTotalBytesTransferredOverNetwork):
* Source/WebKit/NetworkProcess/NetworkLoad.cpp:
(WebKit::NetworkLoad::bytesTransferredOverNetworkDelta const):
* Source/WebKit/NetworkProcess/NetworkLoad.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::bufferingTimerFired):
(WebKit::NetworkResourceLoader::sendBuffer):
(WebKit::NetworkResourceLoader::dataReceivedThroughContentFilter):
(WebKit::NetworkResourceLoader::bytesTransferredOverNetworkDelta const):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::didReceiveData):
(WebKit::ServiceWorkerFetchTask::didReceiveDataFromPreloader):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::didReceiveData):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::curlDidReceiveData):
* Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp:
(WebKit::NetworkDataTaskSoup::didRead):
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::didReceiveData):
* Source/WebKit/WebProcess/Network/WebResourceLoader.h:
* Source/WebKit/WebProcess/Network/WebResourceLoader.messages.in:

Canonical link: https://commits.webkit.org/290164@main
